### PR TITLE
Fix BLOOM dtype

### PR DESCRIPTION
**Pull request by Muennighoff**
Issues:
a) The bloom attribute map does not follow the convention of 
```json
common_arg: model_specific_arg
```
& the argument n_embed is there for no apparent reason.

b) Using attributes with defaults in the attribute_map leads to them being overwritten in [setattr](https://github.com/huggingface/transformers/blob/7498db06a1724773ad7185887743afcb8406ee6b/src/transformers/configuration_utils.py#L249) when the respective kwarg is popped [here](https://github.com/huggingface/transformers/blob/7498db06a1724773ad7185887743afcb8406ee6b/src/transformers/configuration_utils.py#L265). For now I just removed the `dtype` argument for `torch_dtype`. I think the best solution for this would be to check the attribute map before popping. I can also implement that instead if someone gives me a heads up.

c) Weights end up in fp32 & not the torch_dtype when loading them in via load_state_dict. I havn't looked into this in more detail, so I just fixed it via casting, which is sufficient for the conversion script I think.

Notes: 
- If we merge this, we need to update all BLOOM configs in the Hub by swapping `dtype` for `torch_dtype` & swapping `n_embed` for `hidden_size`

### CODE REVIEW

**Comment by thomasw21** [comment]
You should just set the `dtype` in the config file

**Comment by Muennighoff** [comment]
Hmm maybe I'm doing sth wrong. I use

`python /gpfsssd/worksf/projects/rech/six/commun/conda/muennighoffmodelconv/lib/python3.8/site-packages/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py --bloom_checkpoint_path /gpfsscratch/rech/six/commun/checkpoints/tr11e-350M-ml/checkpoints/main/global_step659500 --pytorch_dump_folder_path bloomfp16 --bloom_config_file /gpfsscratch/rech/six/commun/commun/experiments/muennighoff/bloomckpt/350M/bloom-350m/config.json --pretraining_tp 1`

where the config is the below cloned from [here](https://huggingface.co/bigscience/bloom-350m)

```json
{
  "apply_residual_connection_post_layernorm": false,
  "attention_dropout": 0.0,
  "attention_softmax_in_fp32": true,
  "bias_dropout_fusion": true,
  "bos_token_id": 1,
  "dtype": "float16",
  "eos_token_id": 2,
  "pad_token_id": 3,
  "unk_token_id": 0,
  "hidden_dropout": 0.0,
  "initializer_range": 0.02,
  "layer_norm_epsilon": 1e-05,
  "masked_softmax_fusion": true,
  "model_type": "bloom",
  "n_embed": 1024,
  "n_inner": null,
  "n_layer": 24,
  "num_attention_heads": 16,
  "offset_alibi": 100,
  "pretraining_tp": 1,
  "seq_length": 2048,
  "skip_bias_add": true,
  "skip_bias_add_qkv": false,
  "transformers_version": "4.20.0.dev0",
  "use_cache": true,
  "vocab_size": 250880
}
```

**Comment by younesbelkada** [comment]
Thanks for the fix! Indeed there seems to be a mistake there, and since the native weights are already in `fp16` / `bfloat16` we did not observed it since the `dtype` can be automatically retrieved here when doing `torch_dtype="auto"`: https://github.com/huggingface/transformers/blob/7498db06a1724773ad7185887743afcb8406ee6b/src/transformers/modeling_utils.py#L196 
I will ask @sgugger to review this since it may break backward compatibility but I think if we change all config files it should be fine

**Review by sgugger — COMMENTED** [review]
`torch_dtype` should not be a parameter of the config subclass. It's automatically set by `save_pretrained` for all configs. So, it shouldn't be documented/treated here in the config and left to the super class.

**Comment by Muennighoff** [comment]
Sure, I will remove `torch_dtype` then entirely from the Bloom config.
Like @younesbelkada said in order to get BLOOM in the correct dtype one will always have to specify `torch_dtype=auto`. The below returns a model in fp32 not the desired bf16:

```python
from transformers import AutoModel, BloomModel
model = BloomModel.from_pretrained("bigscience/bloom-350m")
```

**Comment by sgugger** [comment]
Yes, but that was already the case before this PR.

File: src/transformers/models/bloom/configuration_bloom.py [hunk 1]
```diff
@@ -117,9 +114,7 @@ class BloomConfig(PretrainedConfig):
     keys_to_ignore_at_inference = ["past_key_values"]
     attribute_map = {
         "num_hidden_layers": "n_layer",
-        "n_head": "num_attention_heads",
-        "hidden_size": "n_embed",
```
Comment (sgugger) [inline src/transformers/models/bloom/configuration_bloom.py:121]: Those should not be removed.

File: src/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py [hunk 2]
```diff
@@ -79,7 +79,11 @@ def get_dtype_size(dtype):
 
 
 def convert_bloom_checkpoint_to_pytorch(
-    bloom_checkpoint_path, bloom_config_file, pytorch_dump_folder_path, shard_model, pretraining_tp
+    bloom_checkpoint_path,
+    bloom_config_file,
+    pytorch_dump_folder_path,
+    shard_model,
+    pretraining_tp,
```
Comment (sgugger) [inline src/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py:86]: Why this change?

File: src/transformers/models/bloom/configuration_bloom.py (hunk 1 above)
Comment (Muennighoff) [reply src/transformers/models/bloom/configuration_bloom.py:121]: Why do we need to map `n_embed` to `hidden_size` if the kwarg in `__init___` is called `hidden_size`?
If I'm not mistaken the `"n_head": "num_attention_heads"` attribute map is the wrong way around, so I added it as `"num_attention_heads": "n_head"`

File: src/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py (hunk 2 above)
Comment (Muennighoff) [reply src/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py:86]: applying the `[huggingface formatting](https://huggingface.co/docs/transformers/contributing)` said it's okay, so i guess it's fine?

Comment (sgugger) [reply src/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py:86]: Formatting tests pass on the main branch, so there is no reason for this change.

**Review by sgugger — APPROVED** [review]

**Comment by sgugger** [comment]
Thanks for iterating! I just looked and all the configs online define `n_embed` instead of `hidden_size`. So we will need some backward compatible line in the init to pop `n_embed` from the kwargs and set it as hidden_size to not break everything.

We can't just replace the existing configs as it would break BLOOM with the current version of Transformers. The `dtype` field is different since it wasn't used anywhere.

**Comment by sgugger** [comment]
Yes, we can merge, thanks! We can remove the `dtype` attributes in the configurations but the `n_embed` ones will need to stay for compatibility with older versions of Transformers.

**Comment by Muennighoff** [comment]
> I'm late to chime in. We absolutely need `torch_dtype` in all configs, since some models are fp16 and others are bf16. The user needs to know even if the core doesn't yet use `torch_dtype` directly. It can be used as:
> 
> `AutoModel.from_pretrained(..., torch_dtype=config.torch_dtype)`
> 
> model configs should ideally be created with `save_from_pretrained`, which will do the right thing, but given the size on some of the models, like 176B, we might have to find a workaround for that one.
> 
> And some models are missing crucial configs - hidden size.
> 
> Thank you!

Good point. I think we can add the `torch_dtype` kwarg to the hub configs for information purposes, but it does not change the loading of the models.
I.e. 
```python
from transformers import AutoModel, AutoConfig

model = AutoModel.from_pretrained("Muennighoff/bloom-tiny-random")
model.dtype
```
yields `torch.float32` even though `torch_dtype` is [set](https://huggingface.co/Muennighoff/bloom-tiny-random/blob/main/config.json) to `float16`. I'm not sure if this is a feature or a bug?

So models always have to be loaded with

```python
from transformers import AutoModel, AutoConfig

model = AutoModel.from_pretrained("Muennighoff/bloom-tiny-random", torch_dtype="auto")
model.dtype # torch.bfloat16
```

**Comment by younesbelkada** [comment]
@Muennighoff the behaviour that you are exposing is expected, if you want to load a model with its native dtype you always have to pass `torch_dtype=auto`

**Comment by stas00** [comment]
not **always**, sometimes one needs to instantiate a model w/o having access to weights, so one has to do:
```
AutoModel.from_config(config, torch_dtype=config.torch_dtype)
```

"auto" won't work here.

**Comment by stas00** [comment]
The main reason for `torch_dtype` not being used automatically on load is that most model configs don't have it. So we decided to start populating this meta data and eventually switch to using it. I think we added it a year ago or so.

Otherwise there would be an inconsistent behavior where some models would load in a correct dtype and other won't. So for now it's an explicit way.
